src/app/credExplorer/pagerankTable/aggregation.test.js

```diff
@@ -1,7 +1,422 @@
+      rfz4: {adjacency: adjacencies.rfz4, weight: 0.2},
+      rfk1: {adjacency: adjacencies.rfk1, weight: 0.2},
+      rbk2: {adjacency: adjacencies.rbk2, weight: 0.2},
+      kfr3: {adjacency: adjacencies.kfr3, weight: 0.2},
+    };
+    const scoredConnections = {
+      loop5: {
+        connection: connections.loop5,
+        source: nodes.root,
+        sourceScore: 0.2,
+        connectionScore: 5,
+      },
+      rfz4: {
+        connection: connections.rfz4,
+        source: nodes.zap,
+        sourceScore: 0.2,
+        connectionScore: 4,
+      },
+      kfr3: {
+        connection: connections.kfr3,
+        source: nodes.kif,
+        sourceScore: 0.2,
+        connectionScore: 3,
+      },
+      rbk2: {
+        connection: connections.rbk2,
+        source: nodes.kif,
+        sourceScore: 0.2,
+        connectionScore: 2,
+      },
+      rfk1: {
+        connection: connections.rfk1,
+        source: nodes.kif,
+        sourceScore: 0.2,
+        connectionScore: 1,
+      },
+    };
+    const scoredConnectionsArray = [
+      scoredConnections.loop5,
+      scoredConnections.rfz4,
+      scoredConnections.rfk1,
+      scoredConnections.rbk2,
+      scoredConnections.kfr3,
+    ];
+    return {
+      nodes,
+      edges,
+      nodeTypes,
+      nodeTypesArray,
+      edgeTypes,
+      edgeTypesArray,
+      adjacencies,
+      connections,
+      scoredConnections,
+      scoredConnectionsArray,
+    };
+  }
+  describe("aggregateByNodeType", () => {
+    it("puts every connection in an aggregation", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      let connectionsFound = 0;
+      for (const aggregation of aggregations) {
+        connectionsFound += aggregation.connections.length;
+      }
+      expect(connectionsFound).toEqual(scoredConnectionsArray.length);
+    });
+    it("groups connections by node type", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      for (const aggregation of aggregations) {
+        for (const connection of aggregation.connections) {
+          expect(
+            NodeAddress.hasPrefix(
+              connection.source,
+              aggregation.nodeType.prefix
+            )
+          ).toBe(true);
+        }
+      }
+    });
+    it("when multiple node types match, it uses the most specific one", () => {
+      const {nodeTypes, scoredConnections} = example();
+      const aggregations = aggregateByNodeType(
+        [scoredConnections.rfz4],
+        [nodeTypes.empty, nodeTypes.zap]
+      );
+      expect(aggregations).toHaveLength(1);
+      // It used the zap nodeType rather than the empty nodeType b/c it was more specific
+      expect(aggregations[0].nodeType).toEqual(nodeTypes.zap);
```

Comment:
A more property-based (less example-based)way to do this would be to
check that if a node with source `a` is categorized under prefix `p`,
and `a` has prefix `p'`, then `p'` is a prefix of `p`. The way that
you’ve done it is not unreasonable, but you seemed to be enjoying the
property-based nature overall, so I humbly submit this alternative for
your consideration.

Reply:
I like that this example very clearly exhibits the sub-prefixing behavior, and is very easy to reason about.